electrum/tests/test_lnpeer.py

```diff
@@ -289,38 +289,62 @@ class PeerInTests(Peer):
     DELAY_INC_MSG_PROCESSING_SLEEP = 0  # disable rate-limiting
 
 
-class SquareGraph(NamedTuple):
-    #                A
-    #     high fee /   \ low fee
-    #             B     C
-    #     high fee \   / low fee
-    #                D
-    w_a: MockLNWallet
-    w_b: MockLNWallet
-    w_c: MockLNWallet
-    w_d: MockLNWallet
-    peer_ab: Peer
-    peer_ac: Peer
-    peer_ba: Peer
-    peer_bd: Peer
-    peer_ca: Peer
-    peer_cd: Peer
-    peer_db: Peer
-    peer_dc: Peer
-    chan_ab: Channel
-    chan_ac: Channel
-    chan_ba: Channel
-    chan_bd: Channel
-    chan_ca: Channel
-    chan_cd: Channel
-    chan_db: Channel
-    chan_dc: Channel
-
-    def all_peers(self) -> Iterable[Peer]:
-        return self.peer_ab, self.peer_ac, self.peer_ba, self.peer_bd, self.peer_ca, self.peer_cd, self.peer_db, self.peer_dc
-
-    def all_lnworkers(self) -> Iterable[MockLNWallet]:
-        return self.w_a, self.w_b, self.w_c, self.w_d
+high_fee_channel = {
+   'local_balance_msat': 10 * bitcoin.COIN * 1000 // 2,
+   'remote_balance_msat': 10 * bitcoin.COIN * 1000 // 2,
+   'local_base_fee_msat': 500_000,
+   'local_fee_rate_millionths': 500,
+   'remote_base_fee_msat': 500_000,
+   'remote_fee_rate_millionths': 500,
+}
+
+low_fee_channel = {
+    'local_balance_msat': 10 * bitcoin.COIN * 1000 // 2,
+    'remote_balance_msat': 10 * bitcoin.COIN * 1000 // 2,
+    'local_base_fee_msat': 1_000,
+    'local_fee_rate_millionths': 1,
+    'remote_base_fee_msat': 1_000,
+    'remote_fee_rate_millionths': 1,
+}
+
+GRAPH_DEFINITIONS = {
+    'square_graph': {
+        'alice': {
+            'channels': {
+                # we should use copies of channel definitions if
+                # we want to independently alter them in a test
+                'bob': high_fee_channel.copy(),
+                'carol': low_fee_channel.copy(),
+            },
+        },
+        'bob': {
+            'channels': {
+                'dave': high_fee_channel.copy(),
+            },
+            'config': {
+                'lightning_forward_payments': True,
+                'lightning_forward_trampoline_payments': True,
+            },
```

Comment:
So channels only need to be defined for one of the peers and they will get added to both peers,
but configs are per-peer-specific. Maybe this could be briefly mentioned in a comment.
Looks reasonable.
Note that this construction does not permit asymmetric forwarding fees for a given channel -- but the current tests do not need that so it's fine.

Reply:
True, here I tried to opt for a more compact representation.